File: notebooks/sudoku.py
def _solve_impl(grid, rows, cols, blocks, idx=0):
    x = idx % 9
    y = idx // 9
    if y == 9:
        return True

    if grid[y][x]["locked"]:
        return _solve_impl(grid, rows, cols, blocks, idx+1)

    row = rows[y]
    col = cols[x]
    block = blocks[y//3][x//3]

    for value in range(1, 10):
        if value in row or value in col or value in block:
            continue

        grid[y][x]["value"] = value
        row.add(value)
        col.add(value)
        block.add(value)
        
        if _solve_impl(grid, rows, cols, blocks, idx+1):
            return True
            
        row.remove(value)
        col.remove(value)
        block.remove(value)
    grid[y][x]["value"] = None
    return False


def solve(grid):
    rows = [set() for _ in range(9)]
    cols = [set() for _ in range(9)]
    blocks = [
        [set() for _ in range(3)]
        for _ in range(3)
    ]
    prepared_grid = [[] for _ in range(9)]
    for y, row in enumerate(grid):
        for x, val in enumerate(row):
            prepared_grid[y].append({
                "locked": val is not None,
                "value": val
            })
            if val is not None:
                if val in rows[y] or val in cols[x] or val in blocks[y//3][x//3]:
                    raise ValueError("Grid is invalid")
                rows[y].add(val)
                cols[x].add(val)
                blocks[y//3][x//3].add(val)
    result = _solve_impl(prepared_grid, rows, cols, blocks)
    if not result:
        raise ValueError("Cannot solve this grid")
    return prepared_grid

File: sudoku-web/src/lib/core/sudoku.ts
export type SudokuGrid = number[][]

export class UnsolvableGridError extends Error { }

class Validator {
    private rows: Set<number>[];
    private cols: Set<number>[];
    private blocks: Set<number>[][];

    constructor() {
        this.rows = createSets(9);
        this.cols = createSets(9);
        this.blocks = [
            createSets(3),
            createSets(3),
            createSets(3),
        ];
    }

    tryAdd(val: number, row: number, col: number): boolean {
        if (
            this.rows[row].has(val) ||
            this.cols[col].has(val) ||
            this.blocks[Math.floor(row / 3)][Math.floor(col / 3)].has(val)
        ) {
            return false;
        }

        this.rows[row].add(val);
        this.cols[col].add(val);
        this.blocks[Math.floor(row / 3)][Math.floor(col / 3)].add(val);
        return true;
    }

    remove(val: number, row: number, col: number) {
        this.rows[row].delete(val);
        this.cols[col].delete(val);
        this.blocks[Math.floor(row / 3)][Math.floor(col / 3)].delete(val);
    }
}

const createSets = (n: number): Set<number>[] => {
    const result = [];
    for (let i = 0; i < n; i++) {
        result.push(new Set<number>());
    }
    return result;
}

const solveImpl = (sudokuGrid: SudokuGrid, row: number, col: number, validator: Validator): boolean => {
    if (row === 9) {
        return true;
    }

    const nextRow = col === 8 ? row + 1 : row;
    const nextCol = (col + 1) % 9;

    if (sudokuGrid[row][col] !== 0) {
        return solveImpl(sudokuGrid, nextRow, nextCol, validator);
    }

    for (let check = 1; check < 10; check++) {
        if (!validator.tryAdd(check, row, col)) {
            continue;
        }
        sudokuGrid[row][col] = check;

        if (solveImpl(sudokuGrid, nextRow, nextCol, validator)) {
            return true;
        }

        validator.remove(check, row, col);
    }
    sudokuGrid[row][col] = 0;
    return false;
}

export const solve = (sudokuGrid: SudokuGrid): SudokuGrid => {
    const validator = new Validator();
    const copied = sudokuGrid.map(r => [...r]);

    for (let r = 0; r < 9; r++) {
        for (let c = 0; c < 9; c++) {
            const val = copied[r][c];
            if (val === 0) {
                continue;
            }

            if (!validator.tryAdd(val, r, c)) {
                throw new UnsolvableGridError();
            }

            copied[r][c] = -val;
        }
    }

    if (!solveImpl(copied, 0, 0, validator)) {
        throw new UnsolvableGridError();
    }

    return copied;
}